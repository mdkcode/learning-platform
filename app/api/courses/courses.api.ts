import { collection, getDocs, query, where } from "firebase/firestore";
import { db } from "@/app/configs/firebase/firebase";
import { CourseProps } from "@/app/api/courses/courses.interface";
import { convertTimestampToDateString } from "@/app/configs/utils/format";

export async function getCourseList(queryStr?: string) {
  try {
    const coursesRef = collection(db, "courses");
    const coursesQuery = queryStr
      ? query(
          coursesRef,
          where("name", ">=", queryStr),
          where("name", "<=", queryStr + "\uf8ff")
        )
      : coursesRef;
    const querySnapshot = await getDocs(coursesQuery);
    const courses = querySnapshot.docs.map((doc) => ({
      ...doc.data(),
      id: doc.id,
      created_at: convertTimestampToDateString(doc.data().created_at),
    }));

    return courses as CourseProps[];
  } catch (error) {
    throw error;
  }
}

export async function getUserCourseSubscriptions(userId: string) {
  const subscriptionsRef = collection(db, "subscriptions");
  const subscriptionsQuery = query(
    subscriptionsRef,
    where("userId", "==", userId)
  );
  const querySnapshot = await getDocs(subscriptionsQuery);
  // Get the list of course IDs the user is subscribed to
  const subscribedCourseIds = querySnapshot.docs.map(
    (doc) => doc.data().courseId
  );
  return subscribedCourseIds;
}
