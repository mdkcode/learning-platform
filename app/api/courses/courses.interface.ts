export interface CourseSearchParams {
  searchQuery?: string;
}

export interface CourseProps {
  id: string;
  name: string;
  image_url: string;
  description: string;
  created_at: string;
  duration: number;
}

export interface VideoProps {
  id: string;
  video_url: string;
  title: string;
  duration: number;
  order: number;
}
