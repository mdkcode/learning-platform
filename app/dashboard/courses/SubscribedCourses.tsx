"use client";
import { CourseProps } from "@/app/dashboard/(overview)/CoursesList";

export default function SubscribedCourses() {
  const subscribedCourses = JSON.parse(
    localStorage.getItem("subscriptions") ?? ""
  );

  return (
    <>
      {subscribedCourses && subscribedCourses.length ? (
        subscribedCourses.map(({ id, title }: CourseProps) => (
          <div
            className="w-[400px] mb-4 shadow-md rounded-lg p-5 bg-white cursor-pointer transition-all duration-300 transform 
      hover:scale-105 hover:from-white-200"
            key={id?.toString()}
          >
            <label htmlFor={id?.toString()}>{title}</label>
            <progress
              id="file"
              max="100"
              value="0"
              className="w-full h-2 [&::-webkit-progress-bar]:rounded-lg [&::-webkit-progress-value]:rounded-lg
             [&::-webkit-progress-bar]:bg-slate-300
            [&::-webkit-progress-value]:bg-green-600 [&::-moz-progress-bar]:bg-green-600"
            />
          </div>
        ))
      ) : (
        <p className="text-lg font-bold">No courses added yet..</p>
      )}
    </>
  );
}
