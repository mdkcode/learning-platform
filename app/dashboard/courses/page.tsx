export default function Page() {
  return <div>Courses</div>;
}
