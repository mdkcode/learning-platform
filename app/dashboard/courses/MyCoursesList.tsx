"use client";
import { getUserSubscribedCourses } from "@/app/api/courses/courses.api";
import { Routes } from "@/app/routes/routes";
import Link from "@/app/ui/components/Link/Link";
import { useQuery } from "@tanstack/react-query";
import { useSession } from "next-auth/react";

export const MyCoursesList = () => {
  const { data: session } = useSession();
  const userId = session?.userId;
  const { data: subscribedCourses, isLoading } = useQuery({
    queryKey: ["courses", userId],
    queryFn: () => getUserSubscribedCourses(userId!),
    enabled: !!userId,
  });

  return (
    <>
      <h1 className="text-xl font-bold mb-6">My courses</h1>
      {isLoading && <p>Loading...</p>}
      <div className="flex flex-col max-w-[50%] gap-4">
        {subscribedCourses?.map((course) => (
          <div
            key={course.id}
            className="flex justify-between bg-white p-4 border rounded-lg"
          >
            <h1 className="text-lg">{course.name}</h1>
            <Link
              label="View the course"
              href={`${Routes.COURSES}/${course.id}`}
              dataCy="course-link"
            />
          </div>
        ))}
      </div>
    </>
  );
};
