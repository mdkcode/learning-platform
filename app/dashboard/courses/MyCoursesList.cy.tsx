import { MyCoursesList } from "./MyCoursesList";
import * as nextAuth from "next-auth/react";
import * as reactQuery from "@tanstack/react-query";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

describe("<MyCoursesList />", () => {
  const queryClient = new QueryClient();
  it("should show loading state when data is still being fetched", () => {
    cy.stub(nextAuth, "useSession").returns({
      data: { userId: "user123" },
      status: "authenticated",
    });

    cy.stub(reactQuery, "useQuery").returns({
      data: null,
      isLoading: true,
    });
    cy.mount(
      <QueryClientProvider client={queryClient}>
        <MyCoursesList />
      </QueryClientProvider>
    );
    cy.contains("Loading...").should("exist");
  });

  it("should not render courses if session is invalid", () => {
    cy.stub(nextAuth, "useSession").returns({
      data: null,
      status: "unauthenticated",
    });
    cy.stub(reactQuery, "useQuery").returns({
      data: [],
      isLoading: false,
    });
    cy.mount(
      <QueryClientProvider client={queryClient}>
        <MyCoursesList />
      </QueryClientProvider>
    );
    cy.contains("My courses").should("exist");
    cy.get("a").should("not.exist");
  });

  it("should render courses when session is valid and data is fetched", () => {
    cy.stub(nextAuth, "useSession").returns({
      data: { userId: "user123" },
      status: "authenticated",
    });

    const mockCourses = [
      { id: "1", name: "Course 1" },
      { id: "2", name: "Course 2" },
    ];
    cy.stub(reactQuery, "useQuery").returns({
      data: mockCourses,
      isLoading: false,
    });
    cy.mount(
      <QueryClientProvider client={queryClient}>
        <MyCoursesList />
      </QueryClientProvider>
    );
    cy.contains("Course 1").should("exist");
    cy.contains("Course 2").should("exist");
    cy.get("a").should("have.length", 2);
  });
});
