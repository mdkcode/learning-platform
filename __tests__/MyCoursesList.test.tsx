import { render } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";
import { MyCoursesList } from "@/app/dashboard/courses/MyCoursesList";
import { useSession } from "next-auth/react";
import { useQuery } from "@tanstack/react-query";

vi.mock("next-auth/react", () => ({
  useSession: vi.fn(),
}));

vi.mock("@tanstack/react-query", () => ({
  useQuery: vi.fn(),
}));

vi.mock("@/app/ui/components/Link/Link", () => ({
  default: ({ label, href }: { label: string; href: string }) => (
    <a href={href}>{label}</a>
  ),
}));

describe("MyCoursesList", () => {
  it("matches snapshot with loaded courses", () => {
    (useSession as ReturnType<typeof vi.fn>).mockReturnValue({
      data: { userId: "user-123" },
    });

    (useQuery as ReturnType<typeof vi.fn>).mockReturnValue({
      data: [
        { id: "course-1", name: "React Basics" },
        { id: "course-2", name: "Advanced TypeScript" },
      ],
      isLoading: false,
    });

    const { container } = render(<MyCoursesList />);
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot while loading", () => {
    (useSession as ReturnType<typeof vi.fn>).mockReturnValue({
      data: { userId: "user-123" },
    });

    (useQuery as ReturnType<typeof vi.fn>).mockReturnValue({
      data: undefined,
      isLoading: true,
    });

    const { container } = render(<MyCoursesList />);
    expect(container).toMatchSnapshot();
  });
});
