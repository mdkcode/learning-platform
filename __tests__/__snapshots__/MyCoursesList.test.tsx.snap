// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MyCoursesList > matches snapshot while loading 1`] = `
<div>
  <h1
    class="text-xl font-bold mb-6"
  >
    My courses
  </h1>
  <p>
    Loading...
  </p>
  <div
    class="flex flex-col max-w-[50%] gap-4"
  />
</div>
`;

exports[`MyCoursesList > matches snapshot with loaded courses 1`] = `
<div>
  <h1
    class="text-xl font-bold mb-6"
  >
    My courses
  </h1>
  <div
    class="flex flex-col max-w-[50%] gap-4"
  >
    <div
      class="flex justify-between bg-white p-4 border rounded-lg"
    >
      <h1
        class="text-lg"
      >
        React Basics
      </h1>
      <a
        href="/dashboard/courses/course-1"
      >
        View the course
      </a>
    </div>
    <div
      class="flex justify-between bg-white p-4 border rounded-lg"
    >
      <h1
        class="text-lg"
      >
        Advanced TypeScript
      </h1>
      <a
        href="/dashboard/courses/course-2"
      >
        View the course
      </a>
    </div>
  </div>
</div>
`;
